ee/tabby-db/src/github_oauth_credential.rs

```diff
@@ -0,0 +1,123 @@
+use anyhow::Result;
+use chrono::{DateTime, Utc};
+use rusqlite::{params, OptionalExtension};
+
+use super::DbConn;
+
+pub struct GithubOAuthCredentialDAO {
+    pub id: i32,
+    pub client_id: String,
+    pub client_secret: String,
+    pub created_at: DateTime<Utc>,
+    pub updated_at: DateTime<Utc>,
+}
+
+impl GithubOAuthCredentialDAO {
+    fn from_row(row: &rusqlite::Row<'_>) -> std::result::Result<Self, rusqlite::Error> {
+        Ok(Self {
+            id: row.get(0)?,
+            client_id: row.get(1)?,
+            client_secret: row.get(2)?,
+            created_at: row.get(3)?,
+            updated_at: row.get(4)?,
+        })
+    }
+}
+
+/// db read/write operations for `github_oauth_credential` table
+impl DbConn {
+    pub async fn upsert_github_oauth_credential(
+        &self,
+        client_id: &str,
+        client_secret: &str,
+    ) -> Result<i32> {
+        let client_id = client_id.to_string();
+        let client_secret = client_secret.to_string();
+
+        if let Ok(Some(cred)) = self.get_github_oauth_credential().await {
+            self.conn
+                .call(move |c| {
+                    let mut stmt = c.prepare(
+                        r#"UPDATE github_oauth_credential SET client_id = ?, client_secret = ?, updated_at = datetime('now') WHERE id = ?"#,
+                    )?;
+                    stmt.execute(params![client_id, client_secret, cred.id])?;
+                    Ok(())
+                })
+                .await?;
+
+            return Ok(cred.id);
+        }
+
+        let id = self
+            .conn
+            .call(move |c| {
+                let mut stmt = c.prepare(
+                    r#"INSERT INTO github_oauth_credential (client_id, client_secret) VALUES (?, ?)"#,
+                )?;
+                let row_id = stmt.insert(params![client_id, client_secret])?;
+                Ok(row_id)
+            })
+            .await?;
+
+        Ok(id as i32)
+    }
+
+    pub async fn get_github_oauth_credential(&self) -> Result<Option<GithubOAuthCredentialDAO>> {
+        let token = self
+            .conn
+            .call(|conn| {
+                Ok(conn
+                    .query_row(
+                        r#"SELECT id, client_id, client_secret, created_at, updated_at FROM github_oauth_credential"#,
+                        [],
+                        GithubOAuthCredentialDAO::from_row,
+                    )
+                    .optional())
+            })
+            .await?;
+
+        Ok(token?)
+    }
+
+    pub async fn delete_github_oauth_credential(&self) -> Result<()> {
```

Comment:
instead of deleting, i feel it’s better to have a boolean column in table `active` to mark whether the github oauth is enabled.